pkg/resource/deploy/step_executor.go

```diff
@@ -0,0 +1,301 @@
+// verbatim to the post-step event.
+//
+
+// executeStep executes a single step, returning true if the step execution was successful and
+// false if it was not.
+func (se *stepExecutor) executeStep(workerID int, step Step) error {
+	var payload interface{}
+	events := se.opts.Events
+	if events != nil {
+		var err error
+		payload, err = events.OnResourceStepPre(step)
+		if err != nil {
+			se.log(workerID, "step %v on %v failed pre-resource step: %v", step.Op(), step.URN(), err)
+			return errors.Wrap(err, "pre-step event returned an error")
+		}
+	}
+
+	se.log(workerID, "applying step %v on %v (preview %v)", step.Op(), step.URN(), se.preview)
+	status, err := step.Apply(se.preview)
+	if err == nil {
+		// If we have a state object, and this is a create or update, remember it, as we may need to update it later.
+		if step.Logical() && step.New() != nil {
+			if prior, has := se.pendingNews.Load(step.URN()); has {
+				return errors.Errorf(
+					"resource '%s' registered twice (%s and %s)", step.URN(), prior.(Step).Op(), step.Op())
+			}
+
+			se.pendingNews.Store(step.URN(), step)
+		}
+	}
+
+	if events != nil {
+		if postErr := events.OnResourceStepPost(payload, step, status, err); postErr != nil {
+			se.log(workerID, "step %v on %v failed post-resource step: %v", step.Op(), step.URN(), postErr)
+			return errors.Wrap(postErr, "post-step event returned an error")
+		}
+	}
+
+	if err != nil {
+		se.log(workerID, "step %v on %v failed with an error: %v", step.Op(), step.URN(), err)
+		return errStepApplyFailed
+	}
+
+	return nil
+}
+
+// log is a simple logging helper for the step executor.
+func (se *stepExecutor) log(workerID int, msg string, args ...interface{}) {
+	if logging.V(stepExecutorLogLevel) {
+		message := fmt.Sprintf(msg, args...)
+		logging.V(stepExecutorLogLevel).Infof("StepExecutor worker(%d): %s", workerID, message)
+	}
+}
+
+//
+// The step executor owns a number of goroutines that it considers to be "workers", responsible for
+// executing steps. By default, as we ease into the waters of parallelism, there is at most one worker
+// active.
+//
+// Workers continuously pull from se.incomingChains, executing chains as they are provided to the executor.
+// There are two reasons why a worker would exit:
+//
+//  1. A worker exits if se.ctx is canceled. There are two ways that se.ctx gets canceled: first, if there is
+//     a step error in another worker, it will cancel the context. Second, if the plan executor experiences an
+//     error when generating steps or doing pre or post-step events, it will cancel the context.
+//  2. A worker exits if it experiences an error when running a step.
+//
+
+// worker is the base function for all step executor worker goroutines. It continuously polls for new chains
+// and executes any that it gets from the channel.
+func (se *stepExecutor) worker(workerID int) {
+	se.log(workerID, "worker coming online")
+	se.workers.Add(1)
```

Comment:
In principle this should be done outside the goroutine: it creates a race between the worker goroutines starting and the creating goroutine waiting on the wait group. For example, `WaitForCompletion` may return immediately if the workers don't get a chance to start before it is called.

Reply:
good point!